waiter/src/waiter/auth/oauth/github.clj

```diff
@@ -0,0 +1,88 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.auth.oauth.github
+  (:require [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.data.json :as json]
+            [clojure.string :as str]
+            [full.async :as fa]
+            [qbits.jet.client.http :as http]
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :as ring-codec]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.oauth :as oauth]
+            [waiter.cookie-support :as cookie-support]))
+
+; See https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/identifying-users-for-github-apps/
+(defrecord GitHubOAuthProvider [authenticate-uri-fn client-id client-secret http-client password]
+  oauth/OAuthProvider
+  (display-name [_] "GitHub")
+  (redirect [_ {:keys [query-string] {:keys [path]} :route-params :as request}]
+    (fa/go-try
+      (let [token (oauth/random-str)
+            state (-> {:location (oauth/make-location {:path path
+                                                       :query-string query-string})
+                       :token token}
+                      json/write-str)
+            location (oauth/make-uri "https://github.com/login/oauth/authorize"
+                                     {"client_id" client-id
+                                      "scope" "user:email"
+                                      "redirect_uri" (authenticate-uri-fn request)
+                                      "state" state})]
+        (-> {:headers {"location" location}
+             :status 307}
+            ;; We should store this cookie only long enough to provide a reasonable amount of time
+            ;; for the client to authenticate.
+            (cookie-support/add-encoded-cookie password oauth/OAUTH-COOKIE-NAME token (-> 15 t/minutes t/in-seconds))))))
```

Comment:
Should we extract `(-> 15 t/minutes t/in-seconds)` in a helper function? It is used in both providers and seems like something we should have in `oauth.clj`?